libcaf_core/caf/flow/op/sample.hpp

```diff
@@ -0,0 +1,299 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/cow_vector.hpp"
+#include "caf/flow/observable_decl.hpp"
+#include "caf/flow/observer.hpp"
+#include "caf/flow/op/cold.hpp"
+#include "caf/flow/op/state.hpp"
+#include "caf/flow/step/all.hpp"
+#include "caf/flow/subscription.hpp"
+#include "caf/unit.hpp"
+
+#include <vector>
+
+namespace caf::flow::op {
+
+struct sample_input_t {};
+
+struct sample_emit_t {};
+
+///
+template <class T>
+class sample_sub : public subscription::impl_base {
+public:
+  // -- member types -----------------------------------------------------------
+
+  using input_type = T;
+
+  using output_type = T;
+
+  using select_token_type = int64_t;
+
+  // -- constants --------------------------------------------------------------
+
+  static constexpr size_t val_id = 0;
+
+  static constexpr size_t ctrl_id = 1;
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  sample_sub(coordinator* parent, observer<output_type> out)
+    : parent_(parent), out_(std::move(out)) {
+    // nop
+  }
+
+  // -- properties -------------------------------------------------------------
+
+  coordinator* parent() const noexcept override {
+    return parent_;
+  }
+
+  bool running() const noexcept {
+    return state_ == state::running;
+  }
+
+  const error& err() const noexcept {
+    return err_;
+  }
+
+  bool pending() const noexcept {
+    return buf_.has_value();
+  }
+
+  bool can_emit() const noexcept {
+    return buf_.has_value();
+  }
+
+  // -- callbacks for the parent -----------------------------------------------
+
+  void init(observable<input_type> vals, observable<select_token_type> ctrl) {
+    using val_fwd_t = forwarder<input_type, sample_sub, sample_input_t>;
+    using ctrl_fwd_t = forwarder<select_token_type, sample_sub, sample_emit_t>;
+    auto fwd = parent_->add_child(std::in_place_type<val_fwd_t>, this,
+                                  sample_input_t{});
+    vals.subscribe(fwd->as_observer());
+    // Note: the previous subscribe might call on_error, in which case we don't
+    // need to try to subscribe to the control observable.
+    if (running())
+      ctrl.subscribe(parent_->add_child_hdl(std::in_place_type<ctrl_fwd_t>,
+                                            this, sample_emit_t{}));
+  }
+
+  // -- callbacks for the forwarders -------------------------------------------
+
+  void fwd_on_subscribe(sample_input_t, subscription sub) {
+    if (!running() || value_sub_ || !out_) {
+      sub.cancel();
+      return;
+    }
+    value_sub_ = std::move(sub);
+    value_sub_.request(1);
+  }
+
+  void fwd_on_complete(sample_input_t) {
+    value_sub_.release_later();
+    shutdown();
```

Comment:
I don't think this is a robust solution to call `shutdown` here. The whole thing is driven by the ticks, so why not drop the `shutdown` call here and in `fwd_on_error` below and then check in `fwd_on_next(sample_emit_t, ...)` whether `value_sub_` is invalid? This gives the tick an opportunity to emit one last item before calling it a day and the implementation of `shutdown`state_` can become much simpler as a result as well. It doesn't need to switch on the state.